repositories/review_repository.go

```diff
@@ -23,22 +29,170 @@ func DefaultReviewRepository(db *sqlx.DB) *ReviewRepository {
 	var review models.Review
+	err := rr.Database.Get(&review, `SELECT id,
+											rating, 
+											difficulty, 
+											interest, 
+											courseId, 
+											review_date, 
+											review_text, 
+											professor_uniqname, 
+											semester, 
+											userEmail					
+	       							 FROM reviews WHERE id=?`, id)
+	if err != nil {
+		log.Println("Error in GetReviewById:", err)
+		return &review, err
+	}
+
 	return &review, nil
+}
 
+func (rr *ReviewRepository) PostReview(reviewInput *models.Review) error {
+	_, err := rr.Database.NamedExec(`INSERT INTO reviews 
+									 VALUES (DEFAULT,
+											 :rating, 
+											 :difficulty, 
+											 :interest, 
+											 :courseId, 
+											 :review_date,
+											 :review_text, 
+											 :professor_uniqname, 
+											 :semester, 
+											 :userEmail)`, reviewInput)
+	if err != nil {
+		log.Println("Error in PostReview:", err)
+		return err
+	}
+	return nil
+}
+
+func (rr *ReviewRepository) UpdateReview(reviewInput *models.Review, id uint64) (*models.Review, error) {
+	reviewInput.Id = id
+	_, err := rr.Database.NamedExec(`UPDATE reviews 
+									 SET rating = :rating, 
+										 difficulty = :difficulty, 
+										 interest = :interest, 
+										 courseId = :courseId, 
+										 review_date = :review_date, 
+										 review_text = :review_text, 
+										 professor_uniqname = :professor_uniqname, 
+										 semester = :semester, 
+										 userEmail = :userEmail 
+									 WHERE id = :id`, reviewInput)
+	if err != nil {
+		log.Println("Error in UpdateReview:", err)
+		return reviewInput, err
+	}
+	return reviewInput, nil
 }
 
-func (rr *ReviewRepository) TestDB() {
-	_, err := rr.Database.Exec(`CREATE TABLE IF NOT EXISTS test(
-		name varchar(255),
-		value varchar(255)
-	) ENGINE=InnoDB DEFAULT CHARSET=utf8`)
+func (rr *ReviewRepository) DeleteReview(id uint64) error {
+	_, err := rr.Database.Exec(`DELETE FROM reviews WHERE id=?`, id)
+	if err != nil {
+		log.Println("Error in DeleteReview:", err)
+		return err
+	}
+
+	return nil
+}
 
+func (rr *ReviewRepository) GetRatingByReviewId(reviewId uint64) (*models.RatingCount, error) {
+	var rating models.RatingCount
+	err := rr.Database.Get(&rating, `SELECT reviewId,
+       								 		SUM(CASE WHEN helpful = 1 THEN 1 ELSE 0 END) AS helpfulCount,
+       								 		SUM(CASE WHEN helpful = 0 THEN 1 ELSE 0 END) AS notHelpfulCount
+									 FROM ratings WHERE reviewId = ?`, reviewId)
 	if err != nil {
-		log.Println(err, "Could not create table")
-	} else {
-		log.Println("No error")
+		log.Println("Error in GetRatingByReviewId:", err)
+		return &rating, err
 	}
+
+	return &rating, nil
 }
+
+func (rr *ReviewRepository) UpdateRating(ratingInput *models.UserRating) (*models.UserRating, error) {
+	// First, get all ratings matching a reviewId, because two equivalent hashed and salted emails may differ, so we
```

Comment:
So the thing about bcrypt's hashing function is that it can generate many different hashes out of the same parameters, and I couldn't find a way to get an exhaustive list of all the hashes with a certain string and cost, so I used the built in bcrypt function to compare them. I figured that it's not likely for a review to have so many ratings that looping through them would take too much time.

Reply:
Yeah should be fast enough, if we're noticing that the website is very slow we can reevaluate this. Sounds good to me!